pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+        product entirely from the market place and model the market share which
+        accrues to each product in the adjusted market.
+        """
+        if not hasattr(self, "model"):
+            self.sample()
+        if new_utility_equations is None:
+            new_X, new_F, new_y = self.preprocess_model_data(
+                new_choice_df, self.utility_equations
+            )
+            with self.model:
+                if new_F is None:
+                    pm.set_data({"X": new_X, "y": new_y})
+                else:
+                    pm.set_data({"X": new_X, "W": new_F, "y": new_y})
+                # use the updated values and predict outcomes and probabilities:
+                idata_new_policy = pm.sample_posterior_predictive(
+                    self.idata,
+                    var_names=["p", "likelihood"],
+                    return_inferencedata=True,
+                    extend_inferencedata=False,
+                    random_seed=100,
+                )
+
+            self.intervention_idata = idata_new_policy
+        else:
+            new_X, new_F, new_y = self.preprocess_model_data(
+                new_choice_df, new_utility_equations
+            )
+            new_model = self.make_model(new_X, new_F, new_y)
+            with new_model:
+                idata_new_policy = pm.sample_prior_predictive()
+                idata_new_policy.extend(
+                    pm.sample(
+                        target_accept=0.99,
+                        tune=2000,
+                        idata_kwargs={"log_likelihood": True},
+                        random_seed=101,
+                    )
+                )
+                idata_new_policy.extend(
+                    pm.sample_posterior_predictive(
+                        idata_new_policy, var_names=["p", "likelihood"]
+                    )
+                )
+
+            self.intervention_idata = idata_new_policy
+
+        return idata_new_policy
+
+    @staticmethod
+    def calculate_share_change(idata, new_idata):
+        """Calculate difference in market share due to market intervention."""
+        expected = idata["posterior_predictive"].mean(dim=("chain", "draw", "obs"))["p"]
+        expected_new = new_idata["posterior_predictive"].mean(
+            dim=("chain", "draw", "obs")
+        )["p"]
+        shares_df = pd.DataFrame(
+            {"product": expected["alts"], "policy_share": expected}
+        )
+        shares_df_new = pd.DataFrame(
+            {"product": expected_new["alts"], "new_policy_share": expected_new}
+        )
+        shares_df = shares_df.merge(
+            shares_df_new, left_on="product", right_on="product", how="left"
+        )
+        shares_df.fillna(0, inplace=True)
+        shares_df["relative_change"] = (
+            shares_df["new_policy_share"] - shares_df["policy_share"]
+        ) / shares_df["policy_share"]
+        shares_df.set_index("product", inplace=True)
```

Comment:
What about using https://pandas.pydata.org/docs/reference/api/pandas.eval.html to make these computations easier to read?

Reply:
Tried this, just FYI it works nicely in code but breaks in the testing suite for 3.10

<img width="1030" alt="image" src="https://github.com/user-attachments/assets/0ce0dde0-16e0-45d7-a73c-ea643081c44e" />